Overdrive cleanup and improved testability

demiankatz opened the pull request:
This PR adds a new setting to simulate an Overdrive connection, as well as adding fake Overdrive records, to make it possible to test more of the existing Overdrive code without having a real Overdrive account.

It also fixes some issues uncovered during testing:

- Some SolrOverdrive record driver templates had gotten out of sync with their DefaultRecord equivalents (and one was completely redundant).
- Some code triggered notices/warnings that have now been fixed
- Some comments were inaccurate

This currently only takes things far enough to test the templates changed in #2400. In the future, there could be value in simulating more Overdrive functionality and making the fake data more configurable (it presently uses hard-coded values based on the test data added in this PR). However, I think this is a useful start.

demiankatz committed:
Overdrive fixes and testability.

demiankatz committed:
Add missing file.

demiankatz committed:
More robust error handling.

demiankatz committed:
Style fix.

demiankatz committed:
Add test data.

demiankatz committed:
Fix style.

demiankatz committed:
Fix style and error tolerance.

demiankatz committed:
Add fake data to display Overdrive/MyContent page.

demiankatz committed:
Restore removed logic.

demiankatz committed:
Safer refactoring.

EreMaijala commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -6,25 +6,25 @@ $data = [];
+if ($code == 'od_code_login_for_avail'){
     $loginForAvail = true;
-}elseif($this->result->code == 'od_account_noaccess'){
+} elseif($code == 'od_account_noaccess'){
```
Add a space after elseif and spaces before `{` here and above.

EreMaijala commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -6,25 +6,25 @@ $data = [];
 }
 
-foreach($ids as $id):
+foreach($ids as $id) {
```
Space after foreach

EreMaijala commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -33,14 +33,15 @@ foreach($ids as $id):
   $data[] = $current;
-endforeach;
+}
+$jsonResult->data = new \stdClass;
```
`$jsonResult` could be instantiated here as it's not needed before. But do we need stdClass or could $jsonResult be just an associative array?

EreMaijala commented on module/VuFind/src/VuFind/DigitalContent/FakeOverdriveConnector.php:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    /**
+     * Get Colllection Token
```
triple lll (I know it's copied, but still)

EreMaijala commented on module/VuFind/src/VuFind/DigitalContent/FakeOverdriveConnector.php:
```diff
@@ -0,0 +1,325 @@
+    /**
+     * Get Colllection Token
+     *
+     * Gets the colleciton token for the Overdrive collection. The collection
```
Typo: colleciton (ditto)

EreMaijala replied:
This is still there. on line 131.

demiankatz replied:
Sorry, saw the "ditto" and fixed the "Colllection" typo, then missed the different collection-related typo right below it. Should now be fully fixed. :-)

demiankatz committed:
Merge branch 'dev' into overdrive-test-cleanup

demiankatz committed:
Fix whitespace in templates.

demiankatz committed:
Fix more template whitespace.

demiankatz committed:
Fix typo.

demiankatz committed:
Fix return type; add more test data.

demiankatz committed:
Fix bugs; simplify; add translation.

demiankatz commented:
Thanks, @EreMaijala! I've addressed all the issues you raised. You prompted me to do a more careful test/review of status-full.phtml and I discovered quite a few problems which should now be corrected.

Also, some of the style inconsistencies you noted here are found throughout the theme, so I have opened #2709 to fully normalize them. I'll request a review from you on that one after I've run the full test suite and done a more careful review of my search-and-replace work. :-)

EreMaijala commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -28,20 +25,19 @@ foreach ($ids as $id) {
-                        <div class='holds'><strong>Holds</strong> "
+        $current['full_status'] = "<p><div class='copies'><strong>"
+            . $this->transEsc('od_avail_total') . "</strong> "
+            . $avail->copiesOwned . "</div><div class='avail'><strong>"
```
Is there any chance that $avail contains something else than numbers? Maybe escape these to be sure?

demiankatz replied:
It should always be numeric, but it never hurts to escape just in case. Done now!

demiankatz committed:
Add escaping.

demiankatz committed:
Fix another typo.

EreMaijala reviewed (approved):
Looks good to me now. Thanks for bearing all the nit-pickery!
